contracts/cw-proposal-single/src/contract.rs

```diff
@@ -297,11 +301,29 @@ pub fn execute_vote(
         return Err(ContractError::NotRegistered {});
     }
 
+    let mut previous_ballot = None;
     BALLOTS.update(
         deps.storage,
         (proposal_id, info.sender.clone()),
         |bal| match bal {
-            Some(_) => Err(ContractError::AlreadyVoted {}),
+            Some(current_ballot) => {
+                if prop.allow_revoting {
+                    if current_ballot.vote == vote {
+                        // Don't allow casting the same vote more than
+                        // once. This seems liable to be confusing
+                        // behavior.
+                        Err(ContractError::AlreadyCast {})
```

Comment:
Not sure how useful this error message is? Could just overwrite the previous vote regardless of whether it was the same? No strong opinions here though.

Reply:
That was my initial thought as well. Agree the error is a little arbitrary. Was trying to think about this from the perspective of a user of a frontend. If I voted once, and then immediately cast the same vote and didn't see an error it is hard to tell the difference between the first vote succeeding and the first one failing. Think in some ways this serves as a confirmation.

Also, in a frontend we build for this we will almost certainly not allow casting the same vote more than one.